File: invesdwin-scripting-parent/invesdwin-scripting-julia/invesdwin-scripting-julia-runtime-contract/src/test/java/de/invesdwin/scripting/julia/runtime/contract/callback/ParametersAndReturnsTestCharacter.java
package de.invesdwin.scripting.julia.runtime.contract.callback;

import java.util.ArrayList;
import java.util.List;

import javax.annotation.concurrent.NotThreadSafe;

import org.springframework.core.io.ClassPathResource;

import de.invesdwin.scripting.IScriptTaskEngine;
import de.invesdwin.scripting.IScriptTaskInputs;
import de.invesdwin.scripting.IScriptTaskResults;
import de.invesdwin.scripting.callback.IScriptTaskCallback;
import de.invesdwin.scripting.callback.ReflectiveScriptTaskCallback;
import de.invesdwin.scripting.julia.runtime.contract.AScriptTaskJulia;
import de.invesdwin.scripting.julia.runtime.contract.IScriptTaskRunnerJulia;
import de.invesdwin.util.assertions.Assertions;
import de.invesdwin.util.math.Characters;

@NotThreadSafe
public class ParametersAndReturnsTestCharacter {

    private final IScriptTaskRunnerJulia runner;

    public ParametersAndReturnsTestCharacter(final IScriptTaskRunnerJulia runner) {
        this.runner = runner;
    }

    public void testCharacter() {
        new AScriptTaskJulia<Void>() {

            @Override
            public IScriptTaskCallback getCallback() {
                final ParametersAndReturnsTestCharacterCallback callback = new ParametersAndReturnsTestCharacterCallback();
                return new ReflectiveScriptTaskCallback(callback);
            }

            @Override
            public void populateInputs(final IScriptTaskInputs inputs) {}

            @Override
            public void executeScript(final IScriptTaskEngine engine) {
                engine.eval(new ClassPathResource(ParametersAndReturnsTestCharacter.class.getSimpleName() + ".jl",
                        ParametersAndReturnsTestCharacter.class));
            }

            @Override
            public Void extractResults(final IScriptTaskResults results) {
                return null;
            }
        }.run(runner);
    }

    public static class ParametersAndReturnsTestCharacterCallback {

        //putCharacter
        private final char putCharacter;

        //putCharacterVector
        private final char[] putCharacterVector;

        //putCharacterVectorAsList
        private final List<Character> putCharacterVectorAsList;

        //putCharacterMatrix
        private final char[][] putCharacterMatrix;

        //putCharacterMatrixAsList
        private final List<List<Character>> putCharacterMatrixAsList;

        public ParametersAndReturnsTestCharacterCallback() {
            //putCharacter
            this.putCharacter = 'a';

            //putCharacterVector
            this.putCharacterVector = new char[3];
            for (int i = 0; i < putCharacterVector.length; i++) {
                putCharacterVector[i] = Characters.checkedCast('A' + i);
            }

            //putCharacterVectorAsList
            this.putCharacterVectorAsList = Characters.asList(putCharacterVector);

            //putCharacterMatrix
            this.putCharacterMatrix = new char[4][];
            for (int i = 0; i < putCharacterMatrix.length; i++) {
                final char[] vector = new char[3];
                for (int j = 0; j < vector.length; j++) {
                    vector[j] = Characters.checkedCast('A' + i + j);
                }
                putCharacterMatrix[i] = vector;
            }

            //putCharacterMatrixAsList
            this.putCharacterMatrixAsList = new ArrayList<List<Character>>(putCharacterMatrix.length);
            for (final char[] vector : putCharacterMatrix) {
                putCharacterMatrixAsList.add(Characters.asList(vector));
            }
        }

        public char getCharacter() {
            return putCharacter;
        }

        public void setCharacter(final char putCharacter) {
            Assertions.assertThat(this.putCharacter).isEqualTo(putCharacter);
        }

        public char[] getCharacterVector() {
            return putCharacterVector;
        }

        public void setCharacterVector(final char[] putCharacterVector) {
            Assertions.assertThat(this.putCharacterVector).isEqualTo(putCharacterVector);
        }

        public List<Character> getCharacterVectorAsList() {
            return putCharacterVectorAsList;
        }

        public void setCharacterVectorAsList(final List<Character> putCharacterVectorAsList) {
            Assertions.assertThat(this.putCharacterVectorAsList).isEqualTo(putCharacterVectorAsList);
        }

        public char[][] getCharacterMatrix() {
            return putCharacterMatrix;
        }

        public void setCharacterMatrix(final char[][] putCharacterMatrix) {
            Assertions.assertThat(this.putCharacterMatrix).isEqualTo(putCharacterMatrix);
        }

        public List<List<Character>> getCharacterMatrixAsList() {
            return putCharacterMatrixAsList;
        }

        public void setCharacterMatrixAsList(final List<List<Character>> putCharacterMatrixAsList) {
            Assertions.assertThat(this.putCharacterMatrixAsList).isEqualTo(putCharacterMatrixAsList);
        }

    }

}

File: invesdwin-scripting-parent/invesdwin-scripting-runtime-lua/src/test/java/de/invesdwin/scripting/runtime/lua/tests/callback/ParametersAndReturnsTestCharacter.lua
arraysToString = java.method(java.import('de.invesdwin.util.collections.Arrays'),'toString','java.lang.Object')

print("getCharacter")
if(getCharacter ~= nil) then
	error("getCharacter already defined!")
end
getCharacter = callback("getCharacter")
getCharacterType = type(getCharacter)
print(getCharacterType)
print(getCharacter)
if(getCharacterType ~= "string") then
	error("getCharacter not Character!")
end
callback("setCharacter", getCharacter)

print("getCharacterVector")
if(getCharacterVector ~= nil) then
	error("getCharacterVector already defined!")
end
getCharacterVector = callback("getCharacterVector")
getCharacterVectorType = type(getCharacterVector[1])
print(getCharacterVectorType)
print(arraysToString(getCharacterVector))
if(getCharacterVectorType ~= "string") then
	error("getCharacterVector not Character!")
end
callback("setCharacterVector", getCharacterVector)

print("getCharacterVectorAsList")
if(getCharacterVectorAsList ~= nil) then
	error("getCharacterVectorAsList already defined!")
end
getCharacterVectorAsList = callback("getCharacterVectorAsList")
getCharacterVectorAsListType = type(getCharacterVectorAsList[1])
print(getCharacterVectorAsListType)
print(arraysToString(getCharacterVectorAsList))
if(getCharacterVectorAsListType ~= "string") then
	error("getCharacterVectorAsList not Character!")
end
callback("setCharacterVectorAsList", getCharacterVectorAsList)

print("getCharacterMatrix")
if(getCharacterMatrix ~= nil) then
	error("getCharacterMatrix already defined!")
end
getCharacterMatrix = callback("getCharacterMatrix")
getCharacterMatrixType = type(getCharacterMatrix[1][1])
print(getCharacterMatrixType)
print(arraysToString(getCharacterMatrix))
if(getCharacterMatrixType ~= "string") then
	error("getCharacterMatrix not Character!")
end
callback("setCharacterMatrix", getCharacterMatrix)

print("getCharacterMatrixAsList")
if(getCharacterMatrixAsList ~= nil) then
	error("getCharacterMatrixAsList already defined!")
end
getCharacterMatrixAsList = callback("getCharacterMatrixAsList")
getCharacterMatrixAsListType = type(getCharacterMatrixAsList[1][1])
print(getCharacterMatrixAsListType)
print(arraysToString(getCharacterMatrixAsList))
if(getCharacterMatrixAsListType ~= "string") then
	error("getCharacterMatrixAsList not Character!")
end
callback("setCharacterMatrixAsList", getCharacterMatrixAsList)